sway-core/src/declaration_engine/declaration_engine.rs

```diff
@@ -76,17 +77,69 @@ impl DeclarationEngine {
         self.slab.get(*index).expect_function(span)
     }
 
-    fn de_add_monomorphized_function_copy(
+    pub(crate) fn de_get_monomorphized_function(
         &self,
         original_id: DeclarationId,
-        new_copy: TypedFunctionDeclaration,
-    ) {
-        let span = new_copy.span();
-        let new_id = DeclarationId::new(
-            self.slab.insert(DeclarationWrapper::Function(new_copy)),
-            span,
+        type_arguments: &Vec<TypeArgument>,
+    ) -> Option<TypedFunctionDeclaration> {
+        for copy in self
+            .de_get_monomorphized_copies(original_id)
+            .iter()
+            .cloned()
+        {
+            let monomorphized_fn_decl = copy.expect_function(&Span::dummy()).unwrap();
+            if &monomorphized_fn_decl.type_parameters == type_arguments {
```

Comment:
I think that we may need to think more deeply about the logic of this line. In general, most generic functions are called without type arguments, so most of the time this check will fail, missing valid cache opportunities.